rageval/metrics/_answer_bleu.py

```diff
@@ -0,0 +1,162 @@
+
+import datasets
+from datasets import Dataset
+import re
+
+from rageval.metrics import Metric, add_attribute
+
+
+_DESCRIPTION = """\
+BLEU (Bilingual Evaluation Understudy) is an algorithm for evaluating the quality of text which has been machine-translated from one natural language to another.
+Scores are calculated for individual translated segments—generally sentences—by comparing them with a set of good quality reference translations.
+Those scores are then averaged over the whole corpus to reach an estimate of the translation's overall quality.
+Neither intelligibility nor grammatical correctness are not taken into account.
+
+For details, see the paper: http://www.aclweb.org/anthology/P02-1040.pdf
+"""
+
+_KWARGS_DESCRIPTION = """\
+Args:
+    name : str
+    batch_size : int, Batch size for openai completion.
+
+Optional Args:
+    None
+
+Functions:
+    _clean: clean special word in sentence.
+    _compute_single: compute bleu score for single prediction with its references
+
+Examples:
+    >>> from datasets import Dataset
+    >>> import rageval as rl
+    >>> sample = {
+    ...     "answers": [
+    ...         "Hello, world!",
+    ...         "I am a metric named bleu."
+    ...     ],
+    ...     "gt_answers": [
+    ...         [
+    ...             "Hello, my world!",
+    ...         ],
+    ...         [
+    ...             "I am a metric named bleu.",
+    ...             "I am bleu metric.",
+    ...         ]
+    ...     ],
+    ... }
+    >>> dataset = Dataset.from_dict(sample)
+    >>> metric = rl.metrics.AnswerBleuScore()
+    >>> metric.mtype
+    'AnswerCorrectness'
+    >>> s, ds = metric.compute(dataset, batch_size=1)
+    >>> assert 0 <= s <= 1
+    >>> type(ds)
+    <class 'datasets.arrow_dataset.Dataset'>
+
+"""
+
+
+_CITATION = """\
+@misc{Kishore2002bleu,
+      title={Bleu: a method for automatic evaluation of machine translation},
+      author={Kishore Papineni and Salim Roukos and Todd Ward and Wei-Jing Zhu},
+      year={2002},
+      page={311-318},
+      primaryClass={cs.CL}
+}
+"""
+
+
+@dataclass
+@add_attribute('mtype', 'AnswerCorrectness')
+@datasets.utils.file_utils.add_start_docstrings(_DESCRIPTION, _KWARGS_DESCRIPTION)
+class AnswerBleuScore(Metric):
+    """Bleu score computing with good quality reference."""
+
+    name = "answer_bleu"
+
+    ALIAS = ['answer_bleu']
+
+    def __init__(self):
+        """Explicitly initialize the AnswerBleuScore to ensure all parent class initialized."""
+        super().__init__()
+        self._required_columns = ['answers', 'gt_answers']
+
+    def __repr__(self) -> str:
+        """:return: Formatted string representation of the metric."""
+        return f"{self.ALIAS[0]}"
+
+    def _info(self):
+        return datasets.MetricInfo(
+            description=_DESCRIPTION,
+            inputs_description=_KWARGS_DESCRIPTION,
+            citation=_CITATION,
+            homepage="",
+            features=datasets.Features(
```

Comment:
The feature should be like this? as in the implementation in huggingface?
```
    features=datasets.Features(
                {
                    "predictions": datasets.Sequence(datasets.Value("string", id="token"), id="sequence"),
                    "references": datasets.Sequence(
                        datasets.Sequence(datasets.Value("string", id="token"), id="sequence"), id="references"
                      ),
                }
       )

Reply:
In datasets library, the input of bleu metric must be token like ["hello", "there", "general", "kenobi"]
However, I added a tokenizer which split the sentence by whitespace, thus, the input can be strings which will better fit our framework.
by this reason, I think the features do not have to be the same with huggingface.

Reply:
ok